### PR TITLE
Blocks: Ensure that metadata registered on the server for core block is preserved on the client

### PR DESCRIPTION
**Pull request by gziolo**
## Description

Extracted from #29095.
Alternative to https://github.com/WordPress/gutenberg/pull/28829.

Props to @gwwar for the proposed solution.

This PR resolves the issue where metadata registered on the server from `block.json` would get overridden by the same data registered again on the client. The fix proposed ensures that for all blocks that have some metadata present, we skip all other attempts to pass the same metadata. `unstable__bootstrapServerSideBlockDefinitions` was never meant to stay for so long, but here we are. It isn't part of public API so we shouldn't be worried about backward compatibility that much.

It's a blocker for #29095.

It also was caught by @audrasjb when compiling the dev note for new filters added for `register_block_type_from_metadata`. I included e2e tests that ensure that one of the new filters is properly propagated to the client.

## How has this been tested?

There is a new e2e test added that ensures that the updated logic works correctly:

```bash
npm run test-e2e packages/e2e-tests/specs/editor/plugins/register-block-type-hooks.test.js
````

## Types of changes

Bug fix (non-breaking change which fixes an issue). Should be backported to WordPress 5.7. /cc @noisysocks.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] I've tested my changes with keyboard and screen readers. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate. 
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR.

### CODE REVIEW

File: packages/e2e-tests/plugins/register-block-type-hooks.php [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+ * @return array Filtered metadata for registering a block type.
+ */
+function gutenberg_test_block_type_metadata( $metadata ) {
+	if ( 'core/paragraph' !== $metadata[ 'name' ] ) {
+		return $metadata;
+	}
+
+	return array_merge(
+		$metadata,
+		array( 'category' => 'widgets' ),
```
Comment (gziolo) [inline packages/e2e-tests/plugins/register-block-type-hooks.php:24]: I wanted to use `title` but it is still defined on the client outside of `block.json`. It should change in the future.

File: packages/blocks/src/api/registration.js [hunk 2]
```diff
@@ -160,6 +160,11 @@ export const serverSideBlockDefinitions = {};
 // eslint-disable-next-line camelcase
 export function unstable__bootstrapServerSideBlockDefinitions( definitions ) {
 	for ( const blockName of Object.keys( definitions ) ) {
+		// Don't overwrite if already set. It covers the case when metadata
+		// was initialized from the server.
+		if ( serverSideBlockDefinitions[ blockName ] ) {
```
Comment (gwwar) [inline packages/blocks/src/api/registration.js:165]: What's a little fragile here is depending on the load order remaining the same, but I think the E2E should catch the behavior change, if for some reason in the future the client file was read first.

**Review by gwwar — APPROVED** [review]
This tests well for me @gziolo! Thanks for pulling this change out into its own PR 💖

File: packages/blocks/src/api/registration.js (hunk 2 above)
Comment (gziolo) [reply packages/blocks/src/api/registration.js:165]: Right, I didn't think about it before. Looking at how it's processed in WordPress core, it should be reliable:

https://github.com/WordPress/wordpress-develop/blob/d271a206dbc87421b221edb2f7acff6ea523fb5d/src/wp-admin/edit-form-blocks.php#L117-L120

It's glued to `wp-block` script handle, that `wp-block-library` script defines as a dependency.

**Comment by gwwar** [comment]
We need to revert this one @glendaviesnz found that this is wrapping blocks in additional divs

**Comment by talldan** [comment]
Removing the backport label as there's a revert for this PR - #29279

**Comment by ellatrix** [comment]
This PR is causing all blocks to wrapped in a `div` and the cover block is no longer looking correct. Probably we accidentally reverted back to API v1 for all blocks?

**Comment by gziolo** [comment]
Yes, it's because `apiVersion` wasn't exposed from the server. It was fixed in WordPress core and is going to be shipped with v5.7 but it requires a shim for older versions of WordPress that Gutenberg supports. I will create another PR tomorrow latest. We need to read `apiVersion` from `block.json` on the client if not passed from the server.

I checked all `block.json` files and I confirmed that all other fields are properly exposed from the server already so this is the last missing backward compatibility concern. @gwwar, if we end up exposing `variations` on the server, the same issue will pop up.

**Comment by ockham** [comment]
Can we add an automated test (e2e, snapshot, or otherwise) to guard against this? What would be the most reliable/generalized/minimal way of covering this? #bernieTestNag

**Comment by ellatrix** [comment]
@ockham I'll make a simple test

**Pull request merged**